README.md

[![Travis-CI Build Status](https://travis-ci.org/magosil86/getspres.svg?branch=master)](https://travis-ci.org/magosil86/getspres)
[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/magosil86/getspres/master/LICENSE)
[![GitHub issues](https://img.shields.io/github/issues/magosil86/getspres.svg)](https://github.com/magosil86/getspres/issues)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/getspres)](https://cran.r-project.org/package=getspres)
[![Coverage status](https://codecov.io/gh/magosil86/getspres/branch/master/graph/badge.svg)](https://codecov.io/github/magosil86/getspres?branch=master)
[![CRAN_Logs_Rstudio](https://cranlogs.r-pkg.org/badges/grand-total/getspres)](http://cran.rstudio.com/web/packages/getspres/index.html) 
<!-- []() <img src="https://user-images.githubusercontent.com/8364031/46570331-0c8f0180-c952-11e8-923d-536448190113.gif" align="right" /> -->


## _getspres_ : _SPRE_ Statistics for Exploring Heterogeneity in Meta-Analysis


One useful way of identifying overly influential outlier studies in meta-analyses and 
their direction of effect is through the calculation of _SPRE_ (standardised predicted random-effects) 
statistics. _SPRE_ statistics are precision-weighted residuals that capture the direction
and extent to which genetic effects of different studies in a meta-analysis deviate from
the average genetic effect at a variant of interest. Positive outliers have the potential 
to inflate average genetic effects in a meta-analysis whilst negative outliers 
might lower or change the direction of effect.


**_getspres_** facilitates calculation of _SPRE_ statistics in R and provides forest plots 
that show corresponding _SPRE_ statistic values for participating studies in meta-analyses. 

An advantage of using the **_getspres_** package is that it provides a quantitative and 
visual view of heterogeneity at individual genetic variants in meta-analyses.

**Zoom in on the web browser to get a better view of the images**
# []() <img src="image-spres-forestplots.png" align="right"  />


### How are spres calculated?

Consider a GWAS meta-analysis (P), with S GWAS studies (s = 1,2,3,⋯,S) and 
V independently associated variants (v = 1,2,3,⋯,V). Data for each variant are analyzed using a 
random-effects model to estimate the average genetic effect and partition the variability 
in study effect-sizes into random sampling and heterogeneity components. Then, the standardized 
predicted random effect (SPRE) for the vth variant in the sth study is calculated as:


# []() <img src="spre_eqn.png" align="centre" width="500" height="500"/><br>

See the following references for more details:

Harbord, R. M., & Higgins, J. P. T. (2008). Meta-regression in Stata. Stata Journal 8: 493‚Äì519.

Magosi LE, Goel A, Hopewell JC, Farrall M, on behalf of the CARDIoGRAMplusC4D Consortium (2017) 
Identifying systematic heterogeneity patterns in genetic association meta-analysis studies. 
PLoS Genet 13(5): e1006755. https://doi.org/10.1371/journal.pgen.1006755.

---

### Installation

```{r}
# To install the release version from CRAN:
install.packages("getspres")

# Load libraries
library(getspres)  # for calculating SPRE statistics


# To install the development version from GitHub:

# install devtools
install.packages("devtools")

# install getspres
library(devtools)
devtools::install_github("magosil86/getspres")

# Load libraries
library(getspres)  # for calculating SPRE statistics
```

### Usage

Load the getspres package in your current R session, and try some examples in the [example workflow](https://htmlpreview.github.io/?https://github.com/magosil86/getspres/blob/master/vignettes/getspres-tutorial.html)

```{r}
# Load libraries
library(getspres)  # for calculating SPRE statistics
```
For an overview of available functions in **_getspres_**, type `?getspres` and `?plotspres` at the R prompt.



### Getting help

To suggest a new feature, report a bug or ask for help, please provide a reproducible example 
at: https://github.com/magosil86/getspres/issues. Also see [reprex](https://reprex.tidyverse.org/) 
to learn more about generating reproducible examples.


### References

Lerato E Magosi, Anuj Goel, Jemma C Hopewell, Martin Farrall, on behalf of the CARDIoGRAMplusC4D 
Consortium, Identifying small-effect genetic associations overlooked by the conventional 
fixed-effect model in a large-scale meta-analysis of coronary artery disease, Bioinformatics, , btz590, 
https://doi-org.ezp-prod1.hul.harvard.edu/10.1093/bioinformatics/btz590

Harbord, R. M., & Higgins, J. P. T. (2008). Meta-regression in Stata. Stata Journal 8: 493‚Äì519.

Wolfgang Viechtbauer (2010). Conducting meta-analyses in R with the
metafor package. Journal of Statistical Software, 36(3), 1-48. URL
http://www.jstatsoft.org/v36/i03/.

Magosi LE, Goel A, Hopewell JC, Farrall M, on behalf of the CARDIoGRAMplusC4D Consortium (2017) 
Identifying systematic heterogeneity patterns in genetic association meta-analysis studies. 
PLoS Genet 13(5): e1006755. https://doi.org/10.1371/journal.pgen.1006755.


### Authors

<a >Lerato E. Magosi</a> <small class="roles"> </small> <a  href="http://orcid.org/0000-0002-3388-9892" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a>, <a >Jemma C. Hopewell</a> <small class="roles"> </small> and 
<a >Martin Farrall</a> <small class="roles"> </small> <a href="http://orcid.org/0000-0003-4564-2165" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a>


### Maintainer

Lerato E. Magosi lmagosi@well.ox.ac.uk or magosil86@gmail.com

### Citation

Lerato E Magosi, Jemma C Hopewell and Martin Farrall (2018). getspres:
SPRE Statistics for Exploring Heterogeneity in Meta-Analysis. R
package version 0.1.0.9000. https://magosil86.github.io/getspres/
  
### Code of conduct

Contributions are welcome. Please observe the [Contributor Code of Conduct](https://github.com/magosil86/getspres/blob/master/CONDUCT.md) when participating in this project.

### License

[MIT](https://opensource.org/licenses/mit-license.php) + file [License](https://github.com/magosil86/getspres/blob/master/LICENSE).


NEWS.md
# getspres 0.2.0

## Minor improvements

* Added rmarkdown to suggested dependencies


# getspres 0.1.1

## Minor improvements

* Maintenance release for compatibility with R-devel, 
  replaced instances of T with TRUE and F with FALSE
  
  
# getspres 0.1.0

tests/testthat.R
library(testthat)
library(getspres)

test_check("getspres")


tests/testthat/test_getspres_analysis_examples_using_3_lead_variants_in_heartgenes214.R
### library(getspres); library(testthat); Sys.setenv(NOT_CRAN="true")

### see also:
###     https://magosil86.github.io/getspres/
###
###     Magosi LE, Goel A, Hopewell JC, Farrall M, on behalf of the CARDIoGRAMplusC4D Consortium (2018) 
###     Identifying small-effect genetic associations overlooked by the conventional fixed-effect model 
###     in a large-scale meta-analysis of coronary artery disease. 
###     To submit for publication.


context("Checking getspres example: Calculate SPRE statistics for 3 variants in the heartgenes214 dataset ")


test_that("correct SPRE statistics are calculated.", {


	### Test may take longer than 1 min hence test will be run locally rather than on CRAN
	skip_on_cran()

	### Load libraries

	   
	### Load heartgenes214 dataset
	### ?heartgenes214 to view the dataset documentation
	data(heartgenes214, package="getspres")


	### Calculating SPRE statistics for 3 variants in heartgenes214

	### Note: Would have liked to calculate SPRE statistics and generate forest plots 
	###       for variants: "rs16856093", "rs10890302" and "rs10139550" as they were 
	###       presented as figures 3 - 5 in the manuscript draft referred to above. 
	###       However, only "rs10139550" has data available in heartgenes214 hence 
	###       selected two other variants ("rs10168194" and "rs11191416") as examples.

	heartgenes3 <- subset(heartgenes214, 
		variants %in% c("rs10139550", "rs10168194", "rs11191416"))
		
	getspres_results <- getspres(beta_in = heartgenes3$beta_flipped, 
								   se_in = heartgenes3$gcse, 
						  study_names_in = heartgenes3$studies, 
						variant_names_in = heartgenes3$variants)


	### Retrieve number of studies and variants
	num_variants <- getspres_results$number_variants
	num_studies <- getspres_results$number_studies

	### Retrieve SPRE dataset
	df_spres <- getspres_results$spre_dataset


	### checking number of studies and variants
	expect_equal(num_variants, 3)
	expect_equal(num_studies, 48)

	### checking SPRE statistic values for studies 1 - 6 at each of the 3 variants
	expect_equal(round(subset(df_spres[, c("spre")], df_spres$variant_names %in% "rs10168194" & as.numeric(df_spres$study) <= 6), 4), round(c(-0.8196669, -0.2689183, -0.4217485, 0.7116014, -0.6176853, 0.4814332), 4))
	expect_equal(round(subset(df_spres[, c("spre")], df_spres$variant_names %in% "rs10139550" & as.numeric(df_spres$study) <= 6), 4), round(c(-1.7838806, 0.4767771, 0.2405504, 0.8596358, 2.2172400, 1.0915446), 4))
	expect_equal(round(subset(df_spres[, c("spre")], df_spres$variant_names %in% "rs11191416" & as.numeric(df_spres$study) <= 6), 4), round(c(1.0164801, -0.7228475, -0.1810490, 1.7472015, -0.3073316, -0.5469473), 4))


	### Generating a forest plot for variant: "rs10139550" in heartgenes214.
	
	df_rs10139550 <- subset(df_spres, variant_names == "rs10139550")
	
	plotspres_res <- plotspres(beta_in = df_rs10139550$beta, 
								 se_in = df_rs10139550$se, 
						study_names_in = as.character(df_rs10139550$study_names), 
					  variant_names_in = as.character(df_rs10139550$variant_names),
							  spres_in = df_rs10139550$spre,
				   spre_colour_palette = c("multi_colour", "rainbow"),
			  set_studyNOs_as_studyIDs = TRUE,
							  set_xlim = c(-8,8), set_ylim = c(-1.5,51),
								set_at = c(-3.0, -2.5, -2.0, -1.5, -1.0, -0.5, 0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0),
							 save_plot = TRUE)

	### Note: Forest plot for "rs10139550" should resemble figure 5 in the manuscript draft referred to above.


	# Adjust label (i.e. column header) position, also keep plot in graphics window rather
	#     than save as tiff file
	df_rs10139550_3studies <- subset(df_rs10139550, as.numeric(df_rs10139550$study_names) <= 3)

	# Before adjusting label positions
	plotspres_res <- plotspres(beta_in = df_rs10139550_3studies$beta, 
								 se_in = df_rs10139550_3studies$se, 
						study_names_in = as.character(df_rs10139550_3studies$study_names), 
					  variant_names_in = as.character(df_rs10139550_3studies$variant_names),
							  spres_in = df_rs10139550_3studies$spre,
				   spre_colour_palette = c("dual_colour", c("blue","black")),
							 save_plot = FALSE)

	### Note: Forest plot for "rs10139550" should comprise 3 studies.

	# After adjusting label positions
	plotspres_res <- plotspres(beta_in = df_rs10139550_3studies$beta, 
								 se_in = df_rs10139550_3studies$se, 
						study_names_in = as.character(df_rs10139550_3studies$study_names), 
					  variant_names_in = as.character(df_rs10139550_3studies$variant_names),
							  spres_in = df_rs10139550_3studies$spre,
				   spre_colour_palette = c("dual_colour", c("blue","black")),
						 adjust_labels = 1.7, save_plot = FALSE)

	### Note: After adjusting label positions, column headers in the "rs10139550" forest plot 
	###       should now be 2 rows lower.

})


R/driver_compute_spre_statistics.R
# Goal: Generate and export S3 class/methods for getspres and document using roxygen tags
#
#' Exploring Heterogeneity in Meta-Analysis with \emph{SPRE} Statistics.
#' 
#' \code{getspres} computes \emph{SPRE} (standardised predicted random-effects)
#' statistics to identify outlier studies in genetic association meta-analyses
#' which might have undue influence on the average genetic effect leading to 
#' inflated genetic signals.
#'
#' \emph{SPRE} statistics are precision-weighted residuals that summarise the 
#' direction and extent with which observed study effects in a meta-analysis
#' differ from the summary (or average genetic) effect. See the getspres 
#' website for more information, documentation and examples.
#'
#' \code{getspres} takes as input study effect-size estimates and their 
#' corresponding standard errors (i.e. summary data). Study effect estimates 
#' could be in the form of linear regression coefficients or log-transformed
#' regression coefficients (per-allele log odds ratios) from logistic 
#' regression. 
#'
#' \code{getspres} uses inverse-variance weighted meta-analysis models in the
#' \code{metafor} R package to calculate \emph{SPRE} statistics.
#'
#' @seealso \url{https://magosil86.github.io/getspres/} to the visit \code{getspres} website.
#'
#' @aliases getspres getspre spre spres
#'
#' @param beta_in          A numeric vector of study effect-sizes e.g. log odds-ratios. 
#' @param se_in            A numeric vector of standard errors, genomically corrected at study-level.
#' @param study_names_in   A character vector of study names. 
#' @param variant_names_in A character vector of variant names e.g. rsIDs.
#' @param tau2_method      A character scalar, specifying the method that should be used to estimate heterogeneity either through DerSimonian and Laird's moment-based estimate "DL" or restricted maximum likelihood "REML". Note: The REML method uses the iterative Fisher scoring algorithm (step length = 0.5, maximum iterations = 10000) to estimate tau2. Default is "DL".
#' @param verbose_output   An optional boolean to display intermediate output. (Default is \code{FALSE}).
#' @param \dots            other arguments.
#'
#' @return Returns a list containing:
#' \itemize{
#'   \item number_variants A numeric scalar for the number of variants 
#'   \item number_studies  A numeric scalar for the number of studies
#'   \item spre_dataset    A dataframe that is a dataset of computed SPRE statistics and contains the following fields:
#'   \itemize{
#'         \item beta          , study effect-size estimates
#'         \item se            , corresponding standard errors of the study effect-size estimates
#'         \item variant_names , variant names
#'         \item study_names   , study names
#'         \item study         , study numbers
#'         \item snp           , snp numbers
#'         \item tau2          , tau_squared, estimate of amount of between-study variance
#'         \item I2            , I_squared, heterogeneity index (Higgins inconsistency metric)
#'                                 representing proportion of total observed variation due to 
#'                                 between-study variance
#'         \item Q             , Q-statistic (Cochran's Q)
#'         \item xb            , prediction excluding random effects
#'         \item xbse          , standard error of prediction excluding random effects
#'         \item xbu           , predictions including random effects
#'         \item stdxbu        , standard error of prediction (fitted values) 
#'                                 including random effects
#'         \item hat           , leverage a.k.a diagonal elements of the projection hat matrix
#'         \item rawresid      , raw residuals
#'         \item uncondse      , unconditional standard errors
#'         \item spre          , \emph{SPRE} statistics (standardised predicted random effects) i.e. 
#'                                 raw residuals divided by the unconditional standard errors
#'         }
#' }
#'
#'
#' @examples
#' library(getspres)
#'
#' 
#' # Calculate SPRE statistics for a subset of variants in the heartgenes214 dataset.
#' # heartgenes214 is a case-control GWAS meta-analysis of coronary artery disease.
#' # To learn more about the heartgenes214 dataset ?heartgenes214
#'
#' # Calculating SPRE statistics for 3 variants in heartgenes214
#'
#' heartgenes3 <- subset(heartgenes214, 
#'     variants %in% c("rs10139550", "rs10168194", "rs11191416")) 
#'
#' getspres_results <- getspres(beta_in = heartgenes3$beta_flipped, 
#'                                se_in = heartgenes3$gcse, 
#'                       study_names_in = heartgenes3$studies, 
#'                     variant_names_in = heartgenes3$variants)
#'
#'
#' # Explore results generated by the getspres function
#' str(getspres_results)
#'
#' # Retrieve number of studies and variants
#' getspres_results$number_variants
#' getspres_results$number_studies
#'
#' # Retrieve SPRE dataset
#' df_spres <- getspres_results$spre_dataset
#' head(df_spres)
#'
#' # Extract SPREs from SPRE dataset
#' head(spres <- df_spres[, "spre"])
#'
#' \donttest{
#'
#' # Exploring available options in the getspres function:
#' #     1. Estimate heterogeneity using "REML", default is "DL"
#' #     2. Calculate SPRE statistics verbosely
#' getspres_results <- getspres(beta_in = heartgenes3$beta_flipped, 
#'                                se_in = heartgenes3$gcse, 
#'                       study_names_in = heartgenes3$studies, 
#'                     variant_names_in = heartgenes3$variants,
#'                          tau2_method = "REML",
#'                       verbose_output = TRUE)
#'
#' str(getspres_results)
#'
#' }
#'
#' @export
getspres <- function(beta_in, se_in, study_names_in, variant_names_in, ...) UseMethod("getspres")

#' @describeIn getspres Computes \emph{SPRE} statistics in genetic association meta-analyses
#' @export
getspres.default <- function(beta_in, se_in, study_names_in, 
                                  variant_names_in, tau2_method = "DL",
                                  verbose_output = FALSE, ...) {

    # Check whether all required variables are present
    if (missing(beta_in))
        stop("Beta values missing, need to specify a numeric vector of observed study effects.")
 
    if (missing(se_in))
        stop("Standard errors for study effect-size estimates missing, need to specify a numeric vector of standard errors.")

    if (missing(study_names_in))
        stop("Study names missing, need to specify a character vector of study names.")

    if (missing(variant_names_in))
        stop("Variant names missing, need to specify a character vector of variant names.")

    
    # Verify datatypes of required variables
    if (!is.numeric(c(beta_in, se_in))) {

        stop("beta_in and se_in should be of type, numeric.")

    }


    if (!is.character(c(variant_names_in, study_names_in))) {

        stop("variant_names_in and study_names_in should be of type, character.")

    }
    
    compute_spre_results <- compute_spre_statistics(beta_in, se_in, study_names_in, variant_names_in, 
                                              tau2_method, verbose_output)
    
    compute_spre_results$call <- match.call()
    
    class(compute_spre_results) <- "getspres"
    
    compute_spre_results
    
}

#' @export
print.getspres <- function(x, ..., verbose_output = FALSE) {

    cat("Call:\n")
    print(x$call)

    cat("\nNumber of studies:\n")
    print(x$number_studies)

    cat("\nNumber of variants:\n")
    print(x$number_variants)

    cat("\nDataset structure showing SPRE statistics:\n")
    print(str(x$spre_dataset))


}


R/driver_generate_forestplot_incl_spres.R
# Goal: Generate and export S3 class/methods for plotspres and document using roxygen tags
#
#' \code{plotspres} generates forest plots showing \emph{SPRE} statistics.
#'
#'
#' Forest plots showing \emph{SPRE} (standardised predicted random-effects) statistics 
#' can be useful in highlighting overly influential outlier studies with the potential
#' to inflate summary effect estimates in genetic association meta-analyses.
#'
#' \code{plotspres} takes as input \emph{SPRE} statistics, observed study effects 
#' and corresponding standard errors (i.e. summary data). The observed study effects 
#' (i.e. study effect-size estimates) could be association statistics from either 
#' quantitative or binary trait meta-analyses, for instance, linear regression coefficients 
#' might be employed for quantitative traits and log-transformed logistic regression 
#' coefficients (per-allele log odds ratios) used for case-control meta-analyses. 
#' \emph{SPRE} statistics can be calculated using the \code{\link{getspres}} function.
#'
#' \code{plotspres} uses inverse-variance weighted fixed and random-effects 
#'  meta-analysis models in the \code{metafor} R package to generate forestplots.
#' 
#' @seealso \code{\link{getspres}} to calculate \emph{SPRE} statistics and the 
#' \code{\link[metafor:rma.uni]{metafor}} package to explore implementations of fixed and 
#' random-effects meta-analysis models in R. To access more information and examples 
#' visit the getspres website at: \url{https://magosil86.github.io/getspres/}.
#'
#'
#' @aliases plotspres plotspre forestspre spreforest
#'
#' @param beta_in                  A numeric vector of observed study effects e.g. log odds-ratios.
#' @param se_in                    A numeric vector of standard errors, genomically corrected at study-level.
#' @param study_names_in           A character vector of study names.
#' @param variant_names_in         A character vector of variant names e.g. rsIDs.
#' @param spres_in                 A numeric vector of \emph{SPRE} statistics.
#' @param spre_colour_palette      An optional character vector specifying the colour palette that should be used for observed study effects. There are 3 types of colour palettes available, namely: "mono_colour", "dual_colour" and "multi_colour"; with the "dual_colour" palette, observed study effects with negative \emph{SPRE} statistics are coloured differently from those with positive \emph{SPRE} statistics, and with the "multi_colour" palette observed study effects are colored in a gradient according to the \emph{SPRE} statistic values. Default palette option is \code{spre_colour_palette = c("mono_colour", "black")}. 
#' @param set_studyNOs_as_studyIDs An optional boolean specifying whether study numbers should be used as study IDs in the forest plot. Default is \code{FALSE}. 
#' @param set_study_field_width    An optional character vector of format strings, akin to the fmt character vector in the sprintf function. (Default is \code{set_study_field_width = "\%02.0f"}).
#' @param set_cex                  An optional character scalar and symbol expansion factor indicating the percentage by which text and symbols should be scaled relative to the reference; e.g. 1=reference, 1.3 is 30\% larger, 0.3 is 30\% smaller. (Default is \code{cex = 0.66}).
#' @param set_xlim                 An optional numeric vector of length 2 indicating the horizontal limits of the plot region.
#' @param set_ylim                 An optional numeric vector of length 2 indicating the y-axis limits of the plot.
#' @param set_at                   An optional numeric vector indicating position of the x-axis tick marks and corresponding labels.
#' @param tau2_method              An optional character scalar, specifying the method that should be used to estimate heterogeneity either through DerSimonian and Laird's moment-based estimate "DL" or restricted maximum likelihood "REML". Note: The REML method uses the iterative Fisher scoring algorithm (step length = 0.5, maximum iterations = 10000) to estimate tau2. Default is "DL".
#' @param adjust_labels            An optional numeric scalar value that tweaks label (column header) positions. (Default is \code{adjust_labels = 1}).
#' @param save_plot                An optional boolean to save forestplot as a tiff file. Default is \code{TRUE}.
#' @param verbose_output           An optional boolean to display intermediate output. (Default is \code{FALSE}).
#' @param \dots other arguments.
#'         
#' @return Returns a list containing:
#' \itemize{
#'   \item number_variants         A numeric scalar indicating the number of variants 
#'   \item number_studies          A numeric scalar indicating the number of studies
#'   \item fixed_effect_results    A list of fixed-effect meta-analysis results for each variant examined
#'   \item random_effects_results  A list of random-effects meta-analysis results for each variant examined
#'   \item spre_forestplot_dataset A dataframe of the data provided by the user for analysis which contains the following fields:
#'   \itemize{
#'         \item beta            , study effect-size estimates
#'         \item se              , corresponding standard errors of study effect-size estimates
#'         \item variant_names   , variant names
#'         \item study_names     , study names
#'         \item spre            , \emph{SPRE} (standardised predicted random-effects) statistics
#'         \item study_numbers   , study numbers
#'         \item variant_numbers , variant numbers
#'         }
#' }
#'
#' @examples
#' library(getspres)
#'
#' 
#' # Generate a forest plot showing SPRE statistics for variants in heartgenes214.
#' # heartgenes214 is a case-control GWAS meta-analysis of coronary artery disease.
#' # To learn more about the heartgenes214 dataset ?heartgenes214
#'
#' # Calculating SPRE statistics for 3 variants in heartgenes214
#'
#' heartgenes3 <- subset(heartgenes214, 
#'     variants %in% c("rs10139550", "rs10168194", "rs11191416")) 
#'
#' getspres_results <- getspres(beta_in = heartgenes3$beta_flipped, 
#'                                se_in = heartgenes3$gcse, 
#'                       study_names_in = heartgenes3$studies, 
#'                     variant_names_in = heartgenes3$variants)
#'
#' # Explore results generated by the getspres function
#' str(getspres_results)
#'
#' # Retrieve number of studies and variants
#' getspres_results$number_variants
#' getspres_results$number_studies
#'
#' # Retrieve SPRE dataset
#' df_spres <- getspres_results$spre_dataset
#' head(df_spres)
#'
#' # Extract SPREs from SPRE dataset
#' head(spres <- df_spres[, "spre"])
#'
#'
#' # Generating forest plots showing SPREs for variants in heartgenes3
#'
#' # Forest plot with default settings
#' # Tip: To store plots set save_plot = TRUE (useful when generating multiple plots)
#' plotspres_res <- plotspres(beta_in = df_spres$beta, 
#'                              se_in = df_spres$se, 
#'                     study_names_in = as.character(df_spres$study_names), 
#'                   variant_names_in = as.character(df_spres$variant_names),
#'                           spres_in = df_spres$spre,
#'                          save_plot = FALSE)
#'
#' # Explore results generated by the plotspres function
#'
#' # Retrieve number of studies and variants
#' plotspres_res$number_variants
#' plotspres_res$number_studies
#'
#' # Retrieve fixed and random-effects meta-analysis results
#' fixed_effect_res <- plotspres_res$fixed_effect_results
#' random_effects_res <- plotspres_res$random_effects_results
#'
#' # Retrieve dataset that was used to generate forest plots
#' df_plotspres <- plotspres_res$spre_forestplot_dataset
#'
#' \donttest{
#'
#' # Retrieve more detailed meta-analysis output
#' str(plotspres_res)
#' 
#'
#'
#' # Explore available options for plotspres forest plots: 
#' #   1. Colorize study-effect estimates according to SPRE statistic values
#' #   2. Label studies by study number instead of study names
#' #   3. Format study labels (useful when using study numbers as study labels)
#' #   4. Change text size
#' #   5. Adjust x and y axes limits
#' #   6. Change method used to estimate amount of heterogeneity from "DL" to "REML"
#' #   7. Run verbosely to show intermediate results
#' #   8. Adjust label (i.e. column header) positions
#' #   9. Save plot as a tiff file (useful when generating multiple plots)
#'
#' # Colorize study-effect estimates according to SPRE statistic values
#'
#' # Use a dual colour palette for observed study effects so that study effect estimates 
#' #   with negative SPRE statistics are coloured differently from those with positive 
#' #   SPRE statistics.
#' plotspres_res <- plotspres(beta_in = df_spres$beta, 
#'                              se_in = df_spres$se, 
#'                     study_names_in = as.character(df_spres$study_names), 
#'                   variant_names_in = as.character(df_spres$variant_names),
#'                           spres_in = df_spres$spre,
#'                spre_colour_palette = c("dual_colour", c("blue","black")),
#'                          save_plot = FALSE)
#'
#'
#' # Use a multi-colour palette for observed study effects so that study effects estimates
#' #   are colored in a gradient according to SPRE statistic values.
#' #   Available multi-colour palettes:
#' #
#' #       gr_devices_palettes: "rainbow", "cm.colors", "topo.colors", "terrain.colors" 
#' #                            and "heat.colors" 
#' #
#' #       colorspace_hcl_hsv_palettes: "rainbow_hcl", "diverge_hcl", "terrain_hcl", 
#' #                                    "sequential_hcl" and "diverge_hsl"
#' #
#' #       color_ramps_palettes: "matlab.like", "matlab.like2", "magenta2green", 
#' #                             "cyan2yellow", "blue2yellow", "green2red", 
#' #                             "blue2green" and "blue2red"
#'
#' plotspres_res <- plotspres(beta_in = df_spres$beta, 
#'                              se_in = df_spres$se, 
#'                     study_names_in = as.character(df_spres$study_names), 
#'                   variant_names_in = as.character(df_spres$variant_names),
#'                           spres_in = df_spres$spre,
#'                spre_colour_palette = c("multi_colour", "rainbow"),
#'                          save_plot = FALSE)
#'
#' # Exploring other options in the plotspres function.
#' #     Label studies by study number instead of study names (option: set_studyNOs_as_studyIDs)
#' #     Format study labels (option: set_study_field_width)
#' #     Adjust text size (option: set_cex)
#' #     Adjust x and y axes limits (options: set_xlim, set_ylim)
#' #     Change method used to estimate heterogeneity from "DL" to "REML" (option: tau2_method)
#' #     Adjust position of x-axis tick marks (option: set_at)
#' #     Run verbosely (option: verbose_output)
#'
#' df_rs10139550 <- subset(df_spres, variant_names == "rs10139550")
#' plotspres_res <- plotspres(beta_in = df_rs10139550$beta, 
#'                              se_in = df_rs10139550$se, 
#'                     study_names_in = as.character(df_rs10139550$study_names), 
#'                   variant_names_in = as.character(df_rs10139550$variant_names),
#'                           spres_in = df_rs10139550$spre,
#'                spre_colour_palette = c("multi_colour", "matlab.like"),
#'           set_studyNOs_as_studyIDs = TRUE,
#'              set_study_field_width = "%03.0f",
#'                            set_cex = 0.75, set_xlim = c(-2,2), set_ylim = c(-1.5,51),
#'                             set_at = c(-0.6, -0.4, -0.2,  0.0,  0.2,  0.4,  0.6),
#'                        tau2_method = "REML", verbose_output = TRUE,
#'                          save_plot = FALSE)
#'
#' # Adjust label (i.e. column header) position, also keep plot in graphics window rather
#' #     than save as tiff file
#' df_rs10139550_3studies <- subset(df_rs10139550, as.numeric(df_rs10139550$study_names) <= 3)
#'
#' # Before adjusting label positions
#' plotspres_res <- plotspres(beta_in = df_rs10139550_3studies$beta, 
#'                              se_in = df_rs10139550_3studies$se, 
#'                     study_names_in = as.character(df_rs10139550_3studies$study_names), 
#'                   variant_names_in = as.character(df_rs10139550_3studies$variant_names),
#'                           spres_in = df_rs10139550_3studies$spre,
#'                spre_colour_palette = c("dual_colour", c("blue","black")),
#'                          save_plot = FALSE)
#'
#' # After adjusting label positions
#' plotspres_res <- plotspres(beta_in = df_rs10139550_3studies$beta, 
#'                              se_in = df_rs10139550_3studies$se, 
#'                     study_names_in = as.character(df_rs10139550_3studies$study_names), 
#'                   variant_names_in = as.character(df_rs10139550_3studies$variant_names),
#'                           spres_in = df_rs10139550_3studies$spre,
#'                spre_colour_palette = c("dual_colour", c("blue","black")),
#'                      adjust_labels = 1.7, save_plot = FALSE)
#' }
#'
#' @export
plotspres <- function(beta_in, se_in, study_names_in, variant_names_in, spres_in, ...) UseMethod("plotspres")

#' @describeIn plotspres Generates forest plots showing \emph{SPRE} statistics
#' @export
plotspres.default <- function(beta_in, se_in, study_names_in, variant_names_in, 
                              spres_in, spre_colour_palette = c("mono_colour", "black"), 
                              set_studyNOs_as_studyIDs = FALSE, set_study_field_width = "%02.0f", 
                              set_cex = 0.66, set_xlim, set_ylim, set_at, tau2_method = "DL", 
                              adjust_labels = 1, save_plot = TRUE, verbose_output = FALSE, ...) {

    # Check whether all required variables are present
    if (missing(beta_in))
        stop("Beta values missing, need to specify a numeric vector of observed study effects.")
 
    if (missing(se_in))
        stop("Standard errors for study effect-size estimates missing, need to specify a numeric vector of standard errors.")

    if (missing(study_names_in))
        stop("Study names missing, need to specify a character vector of study names.")

    if (missing(variant_names_in))
        stop("Variant names missing, need to specify a character vector of variant names.")
        
    if (missing(spres_in))
        stop("SPRE values missing, need to specify a numeric vector of SPRE statistics.")


    
    # Verify datatypes of required variables
    if (!is.numeric(c(beta_in, se_in, spres_in))) {

        stop("beta_in, se_in and spres_in should be of type, numeric.")

    }


    if (!is.character(c(variant_names_in, study_names_in))) {

        stop("variant_names_in and study_names_in should be of type, character.")

    }
    
    plotspres_results <- generate_spre_forestplot(beta_in, se_in, study_names_in, variant_names_in,
                                                  spres_in, spre_colour_palette, set_studyNOs_as_studyIDs, 
                                                  set_study_field_width, set_cex, set_xlim, set_ylim,
                                                  set_at, tau2_method, adjust_labels, save_plot, 
                                                  verbose_output)
    
    plotspres_results$call <- match.call()
    
    class(plotspres_results) <- "plotspres"
    
    plotspres_results
    
}

#' @export
print.plotspres <- function(x, ..., verbose_output = FALSE) {

    cat("Call:\n")
    print(x$call)

    cat("\nnumber_studies:\n")
    print(x$number_studies)

    cat("\nnumber_variants:\n")
    print(x$number_variants)

    cat("\nDataset structure:\n")
    print(str(x$spre_forestplot_dataset))

}
